src/coreclr/jit/emitxarch.cpp

```diff
@@ -144,6 +144,97 @@ bool emitter::IsDstSrcSrcAVXInstruction(instruction ins)
     return ((CodeGenInterface::instInfo[ins] & INS_Flags_IsDstSrcSrcAVXInstruction) != 0) && IsAVXInstruction(ins);
 }
 
+//------------------------------------------------------------------------
+// IsWriteZFFlags: check if the instruction write the
+//     ZF flag.
+//
+// Arguments:
+//    ins - instruction to test
+//
+// Return Value:
+//    true if instruction writes the ZF flag, false otherwise.
+//
+bool emitter::IsWriteZFFlags(instruction ins)
+{
+    return (CodeGenInterface::instInfo[ins] & INS_FLAGS_WritesZF) != 0;
+}
+
+//------------------------------------------------------------------------
+// IsResetsOCFlags: check if the instruction resets the
+//     OF and CF flag.
+//
+// Arguments:
+//    ins - instruction to test
+//
+// Return Value:
+//    true if instruction resets the OF and CF flag, false otherwise.
+//
+bool emitter::IsResetsOCFlags(instruction ins)
+{
+    return (CodeGenInterface::instInfo[ins] & INS_FLAGS_Resets_CF_OF_Flags) != 0;
+}
+
+//------------------------------------------------------------------------
+// IsFlagsModified: check if the instruction modifies the flags.
+//
+// Arguments:
+//    id - instruction to test
+//
+// Return Value:
+//    true if instruction modified any flag, false otherwise.
+//
+bool emitter::IsFlagsModified(instrDesc* id)
+{
+    instruction ins = id->idIns();
+    insFormat   fmt = id->idInsFmt();
+
+    if (fmt == IF_RRW_SHF)
+    {
+        if (id->idIsLargeCns())
+        {
+            return true;
+        }
+        else if (id->idSmallCns() == 0)
+        {
+            switch (ins)
+            {
+                // If shift-amount for below instructions is 0, then flags are unaffected.
+                case INS_rcl_N:
+                case INS_rcr_N:
+                case INS_rol_N:
+                case INS_ror_N:
+                case INS_shl_N:
+                case INS_shr_N:
+                case INS_sar_N:
+                    return false;
+                default:
+                    return true;
+            }
+        }
+    }
+    else if (fmt == IF_RRW)
+    {
+        switch (ins)
+        {
+            // If shift-amount for below instructions is 0, then flags are unaffected.
+            // So, to be conservative, do not optimize if the instruction has register
+            // as the shift-amount operand.
+            case INS_rcl:
+            case INS_rcr:
+            case INS_rol:
+            case INS_ror:
+            case INS_shl:
+            case INS_shr:
+            case INS_sar:
+                return false;
```

Comment:
I think this might require some careful considerations.

For the purposes of this PR, this is fine and is simply used as a "did the instruction set the flag and therefore a `test` can be skipped" check.

However, for future improvements we may consider more advanced analysis such as:
```
if (!IsFlagsModified(prevInstruction))
{
   // can depend on prevInstruction - 1 state being valid 
}
```

Then, this will become problematic. The latter happens in Clang/MSVC in various places and are one of the reasons that `mulx`, `rorx`, `sarx`, `shlx`, and shrx` are exposed on modern hardware.

If this method were to be used for a "does the previous instruction modify the flags and if not, then the instruction prior to that's state was preserved" check, then this would cause bugs for non-zero inputs.

Reply:
Perhaps we can return some ternary state here representing "yes", "no", and "it depends"?

Reply:
Today, we only track prevInstruction (`emitLastIns`). The way I read your comment, we will need to have access to last couple of instructions (if not more) to make sure the state is valid?

Reply:
Right, it's not something I think we need to be concerned with "today", but rather something I could see being problematic in the future for certain optimizations.

The method implies boolean state but in practice its slightly more nuanced and one of the answers is "maybe", so it may end up confusing callers.

Reply:
Exactly - what are callers supposed to do with "maybe". So for now, I will leave it as-is.

Reply:
> The method implies boolean state but in practice its slightly more nuanced and one of the answers is "maybe", so it may end up confusing callers.

Exactly - what are callers supposed to do with "maybe". So for now, I will leave it as-is.

Reply:
I think that is contextual. In the case of this PR, you care "did the last instruction definitely set the flag I care about" and so you would treat `maybe` as `false`.

In other cases, optimizations may care "did the last instruction definitely not set the flag I care about" and so they would treat it as `true`.

Reply:
Renamed the method to `IsFlagsAlwaysModified`.